gap/weights.gi

```diff
@@ -166,3 +166,453 @@ function(digraph)
   SetEdgeWeightedDigraphTotalWeight(out, total);
   return out;
 end);
+
+#############################################################################
+# 4. Shortest Path
+#############################################################################
+#
+# Three different "shortest path" problems are solved:
+# - All pairs:              DigraphShortestPaths(digraph)
+# - Single source:          DigraphShortestPaths(digraph, source)
+# - Source and destination: DigraphShortestPath (digraph, source, dest)
+#
+# The "all pairs" problem has two algorithms:
+# - Johnson: better for sparse digraphs
+# - Floyd-Warshall: better for dense graphs
+#
+# The "single source" problem has three algorithms:
+# - If "all pairs" is already known, extract information for the given source
+# - Dijkstra: faster, but cannot handle negative weights
+# - Bellman-Ford: slower, but handles negative weights
+#
+# The "source and destination" problem calls the "single source" problem and
+# extracts information for the given destination.
+#
+# Justification and benchmarks are in Raiyan's MSci thesis, Chapter 6.
+#
+
+InstallMethod(EdgeWeightedDigraphShortestPaths,
+"for a digraph with edge weights",
+[IsDigraph and HasEdgeWeights],
+function(digraph)
+  local maxNodes, threshold, digraphVertices, nrVertices, nrEdges;
+
+  digraphVertices := DigraphVertices(digraph);
+  nrVertices := Size(digraphVertices);
+  nrEdges := DigraphNrEdges(digraph);
+
+  maxNodes := nrVertices * (nrVertices - 1);
+
+  # the boundary for performance is edge weight 0.125
+  # so if nr edges for vertices v is less
+  # than total number of edges in a connected
+  # graph we use johnson's algorithm
+  # which performs better on sparse graphs, otherwise
+  # we use floyd warshall algorithm.
+  # This information is gathered from benchmarking tests.
+  threshold := Int(maxNodes / 8);
+  if nrEdges <= threshold then
+    return DIGRAPHS_Edge_Weighted_Johnson(digraph);
+  else
+    return DIGRAPHS_Edge_Weighted_FloydWarshall(digraph);
```

Comment:
Isn't there an issue here if the graph contains a negative cycle? In particular, if it does contain a negative cycle, then shouldn't we use the other algorithm? So, wouldn't it be better to return `fail` from `DIGRAPHS_Edge_Weighted_FloydWarshall` in the case `digraph` contains a negative cycle, and then just run `DIGRAPHS_Edge_Weighted_Johnson` anyway?